File: cpp/medium/3Sum.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
  vector<vector<int>> threeSum(vector<int> &nums) {
    vector<vector<int>> ans;
    int size = nums.size();
    sort(nums.begin(), nums.end());

    for (int i = 0; i < size; i++) {
      if (nums[i] > 0)
        break;
      if (i > 0 && nums[i] == nums[i - 1])
        continue;
      int a = nums[i];
      int l = i + 1, r = size - 1;
      while (l < r) {
        if (nums[l] + nums[r] + a == 0) {
          ans.push_back({a, nums[l], nums[r]});
          l++;
          r--;
          while (l < r && nums[l] == nums[l - 1]) {
            l++;
          }
        } else if ((nums[l] + nums[r] + a) > 0) {
          r--;
        } else {
          l++;
        }
      }
    }
    return ans;
  }
};

File: ts/medium/3Sum.js
const { resolve } = require("path");

class Solution {
	/**
	 * @param {number[]} nums
	 * @return {number[][]}
	 */
	threeSum(nums) {
		nums.sort((a, b) => a - b);
		const result = [];
		for (let idx = 0; idx < nums.length; idx++) {
			let currentNum = nums[idx];
			if (idx > 0 && currentNum == nums[idx - 1])
				continue;

			let [l, r] = [idx + 1, nums.length - 1];
			while (l < r) {
				if (nums[l] + nums[r] + currentNum > 0) {
					r--;
				} else if (nums[l] + nums[r] + currentNum < 0) {
					l++;
				} else {
					result.push([currentNum, nums[l], nums[r]]);
					while (l < r && nums[l] === nums[l + 1]) l++;
					while (l < r && nums[r] === nums[r - 1]) r--;
					l++;
					r--;
				}
			}
		}
		return result;

	}
}
